File: backend/app/routes/photos.py
from datetime import datetime, timezone

import uuid
from flask import Blueprint, jsonify, request

from app.config import PHOTO_BASE_URL
from app.db import get_photos_collection
from app.middleware.auth import require_auth
from app.storage import generate_presigned_put_url

photos_bp = Blueprint("photos", __name__)


def _serialize(doc: dict) -> dict:
    """Convert MongoDB document to a JSON-serializable dict."""
    doc["_id"] = str(doc["_id"])
    return doc


@photos_bp.get("/api/photos")
def get_photos():
    collection = get_photos_collection()
    photos = [_serialize(doc) for doc in collection.find({})]
    return jsonify(photos), 200


@photos_bp.post("/api/photos")
@require_auth
def upload_photo():
    data = request.get_json(silent=True)
    if not data:
        return jsonify({"error": "Invalid request body"}), 400

    title = data.get("title", "").strip()
    if not title:
        return jsonify({"error": "Title is required"}), 400

    description = data.get("description", "")
    aspect_ratio = data.get("aspect_ratio", "")

    photo_id = str(uuid.uuid4())
    s3_key = f"media/photos/{photo_id}/original.jpg"

    try:
        presigned_url = generate_presigned_put_url(s3_key, expires_in=60)
    except RuntimeError as e:
        return jsonify({"error": str(e)}), 500

    now = datetime.now(tz=timezone.utc)
    photo = {
        "id": photo_id,
        "title": title,
        "description": description,
        "aspect_ratio": aspect_ratio,
        "image": {
            "original": f"{PHOTO_BASE_URL}/media/photos/{photo_id}/original.jpg",
            "full": f"{PHOTO_BASE_URL}/media/photos/{photo_id}/full.jpg",
            "400": f"{PHOTO_BASE_URL}/media/photos/{photo_id}/400.jpg",
        },
        "created_at": now,
        "updated_at": now,
    }

    collection = get_photos_collection()
    result = collection.insert_one(photo)

    return jsonify({
        "id": str(result.inserted_id),
        "presigned_url": presigned_url,
        "photo": _serialize(photo),
    }), 201

File: lib/photos.ts
import { PHOTO_BASE_URL } from "@/lib/config.ts";
import { getPhotosCollection } from "@/lib/db.ts";
import { generatePresignedPutUrl } from "@/lib/storage.ts";

export interface PhotoImage {
  original: string;
  full: string;
  "400": string;
}

export interface PhotoDocument {
  id: string;
  title: string;
  description: string;
  aspect_ratio: string;
  image: PhotoImage;
  created_at: Date;
  updated_at: Date;
}

/** A photo as sent to the client: MongoDB's ObjectId rendered as a string. */
export interface Photo
  extends Omit<PhotoDocument, "created_at" | "updated_at"> {
  _id: string;
  created_at: string;
  updated_at: string;
}

export interface NewPhoto {
  title: string;
  description: string;
  aspect_ratio: string;
}

export async function listPhotos(): Promise<Photo[]> {
  const documents = await getPhotosCollection().find({}).toArray();

  return documents.map((doc) => ({
    ...doc,
    _id: doc._id.toString(),
    created_at: doc.created_at.toISOString(),
    updated_at: doc.updated_at.toISOString(),
  }));
}

/**
 * Stores the photo metadata and returns a presigned URL the client uses to
 * upload the original file straight to S3.
 */
export async function createPhoto(input: NewPhoto): Promise<{
  id: string;
  presigned_url: string;
  photo: Photo;
}> {
  const photoId = crypto.randomUUID();
  const s3Key = `media/photos/${photoId}/original.jpg`;
  const presignedUrl = await generatePresignedPutUrl(s3Key, 60);

  const now = new Date();
  const photo: PhotoDocument = {
    id: photoId,
    title: input.title,
    description: input.description,
    aspect_ratio: input.aspect_ratio,
    image: {
      original: `${PHOTO_BASE_URL}/media/photos/${photoId}/original.jpg`,
      full: `${PHOTO_BASE_URL}/media/photos/${photoId}/full.jpg`,
      "400": `${PHOTO_BASE_URL}/media/photos/${photoId}/400.jpg`,
    },
    created_at: now,
    updated_at: now,
  };

  const result = await getPhotosCollection().insertOne(photo);

  return {
    id: result.insertedId.toString(),
    presigned_url: presignedUrl,
    photo: {
      ...photo,
      _id: result.insertedId.toString(),
      created_at: now.toISOString(),
      updated_at: now.toISOString(),
    },
  };
}
